scalafix-reflect/src/main/scala-3/scalafix/internal/reflect/RuleCompiler.scala

```diff
@@ -0,0 +1,71 @@
+package scalafix.internal.reflect
+
+import scalafix.internal.reflect.exception.CompilationException
+
+import java.io.File
+import java.nio.file.Paths
+import scala.meta.io.AbsolutePath
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.Run
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.core.Contexts.NoContext.given_Context
+import dotty.tools.dotc.reporting.StoreReporter
+
+import scala.reflect.io.VirtualDirectory
+import scala.reflect.io.AbstractFile
+import scala.reflect.internal.util.AbstractFileClassLoader
+
+import metaconfig.Configured
+import metaconfig.Input
+
+class RuleCompiler(
+    classpath: String,
+    target: AbstractFile = new VirtualDirectory("(memory)", None)
+) {
+  // we are not configuring any settings!
+  // private val settings = new Settings()
+  // settings.deprecation.value = true // enable detailed deprecation warnings
+  // settings.unchecked.value = true // enable detailed unchecked warnings
+  // settings.outputDirs.setSingleOutput(target)
+  // settings.classpath.value = classpath
+
+  private lazy val reporter: StoreReporter = new StoreReporter()
+  private val compiler: Compiler = new Compiler()
+  private val classLoader: AbstractFileClassLoader =
+    new AbstractFileClassLoader(target, this.getClass.getClassLoader)
+
+  @throws(classOf[CompilationException])
+  def compile(input: Input): Configured[ClassLoader] = {
+    val run: Run = compiler.newRun
+
+    run.compileSources(
+      List(
+        new SourceFile(
+          dotty.tools.io.AbstractFile.getFile(input.path),
+          input.chars
+        )
+      )
+    )
+
+    if (reporter.hasErrors) {
+      val defaultMsg = "Multiple compilation errors occurred for Scala 3, " +
+        "please try to use a CLI version which compiles into Scala 2"
+      val cause = reporter.allErrors(0).getCause()
+
+      if (reporter.errorCount > 1) {
+        val errMsg = reporter.allErrors(0).getMessage()
+        throw new CompilationException(
+          defaultMsg + "\nOriginal message: " + errMsg,
+          cause
+        )
+      }
+      throw new CompilationException(
+        defaultMsg,
+        cause
+      )
+    }
+
+    Configured.Ok(classLoader)
```

Comment:
I am not familiar with dotty, but I wonder how the compiled sources are made available to that classloader since I don't see it passed anywhere during the compilation phase?

Reply:
good point, I'll check this

Reply:
classpath it's now included in the initial context settings for dotty